presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java

```diff
@@ -81,24 +87,74 @@ public static Slice charToVarcharCast(@LiteralParameter("x") Long x, @LiteralPar
     @ScalarOperator(OperatorType.SATURATED_FLOOR_CAST)
     @SqlType("char(y)")
     @LiteralParameters({"x", "y"})
-    // This function returns Char(y) value that is smaller than original Varchar(x) value. However, it is not necessarily the largest
-    // Char(y) value that is smaller than the original Varchar(x) value. This is fine though for usage in TupleDomainTranslator.
-    public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") Long y, @SqlType("varchar(x)") Slice slice)
+    public static Slice varcharToCharSaturatedFloorCast(@LiteralParameter("y") long y, @SqlType("varchar(x)") Slice slice)
     {
-        Slice trimmedSlice = trimSpaces(slice);
-        int trimmedTextLength = countCodePoints(trimmedSlice);
-        int numberOfTrailingSpaces = slice.length() - trimmedSlice.length();
+        List<Integer> codePoints = new ArrayList<>(toCodePoints(slice));
```

Comment:
I haven't looked at the code closely, but is it possible to do this without materializing a list of Integer objects?